Use AJAX api when using Payment Request Buttons with Block-based checkouts

reykjalin opened the pull request:
# Changes proposed in this Pull Request:

Fixes #1524 

The PR changes the block checkout via PRBs such that it uses the pre-existing AJAX APIs to manage the checkout flow, instead of reimplementing the whole flow just for the blocks.

# Testing instructions

Perform various checkout actions through the Payment Request Buttons, e.g.:

- Is a simple purchase successful?
- Do 3DS payments work?
- Are 3DS verificatino failures reported correctly?
- Are card issues reported to the customer, e.g. when a card is declined?
- Are changes to the order information (prior to clicking the PRB) reflected in the payment request dialog?
- Are changes in the payment request dialog reflected in the Cart/Checkout blocks?

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Use AJAX API for successful PRB checkout

This commit is pretty big, but I'm not really sure how to split it into
smaller chunks. For the PRB to function correctly all 3 handlers must be
implemented; shipping options change, shipping address change, and the
payment handler.

That said, 3DS payment function correctly with this commit, but the
Blocks checkout does not. As an example of things the blocks checkout
doesn't currently do:

* Show error messages under certain conditions, e.g. failing the 3DS
  authentication.
* The checkout isn't disabled during payment processing.

Further integration with the Blocks API is needed, but this is a good
first step I think.

Commit:
Use AJAX APIs to checkout with PRBs

We now control the PRBs entirely through AJAX APIs. Errors are reported
correctly and payments are fully functional.

The PRBs and the Blocks are still disconnected though, so any update
made in one (e.g. changing the shipping address) is not reflected in the
other.

For now, the focus is on cleaning up the code, what follows is
implementing the integration with the Blocks.

Commit:
remove obsolete js files

Commit:
move hooks to a separate file

Commit:
add doc comments to the new hooks

Commit:
add comments to the new hooks

Commit:
improve the onclick handler for the PRBs

ricardo commented on client/api/index.js:
```diff
@@ -0,0 +1,166 @@
+	} );
+};
+
+export const createPaymentRequest = ( stripe, cart ) => {
+	const options = {
+		total: cart.order_data.total,
+		currency: cart.order_data.currency,
+		country: cart.order_data.country_code,
+		requestPayerName: true,
+		requestPayerEmail: true,
+		requestPayerPhone:
+			wc_stripe_payment_request_params.checkout.needs_payer_phone,
+		requestShipping: cart.shipping_required ? true : false,
+		displayItems: cart.order_data.displayItems,
+	};
+
+	// Puerto Rico (PR) is the only US territory/possession that's supported by Stripe.
+	// Since it's considered a US state by Stripe, we need to do some special mapping.
+	if ( options.country === 'PR' ) {
+		options.country = 'US';
+	}
+
+	return stripe.paymentRequest( options );
+};
```
It would be great to adapt this function for the product page as well. It would replace [this snippet](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/597cb9b1163c07de7b5af253bf03f87d4d97daf8/assets/js/stripe-payment-request.js#L367) in the shortcode version.

reykjalin replied:
I agree! It'd be great to use all the same functions across both shortcode and block-based checkouts.

Unfortunately that would require moving the shortcode checkout over to the webpack-based build system as opposed to the `uglify` pipeline currently used for the shortcode checkout. And that's in addition to refactoring the code, so while I do think we should do that eventually, I don't think we'll get to it anytime soon.

cc @v18

ricardo replied:
Please ignore these comments about refactoring the shortcode. I added these comments before knowing this PR wasn't intended for that.

Let's keep this simple and merge sooner 👍

ricardo commented on client/api/index.js:
```diff
@@ -0,0 +1,166 @@
+};
+
+export const createOrder = ( sourceEvent, paymentRequestType ) => {
+	const data = getOrderDataFromSourceEvent( sourceEvent, paymentRequestType );
```
I like the way the previous implementation separates the normalization into a separate file. I have kept it in WCPay. We could do the same here.

reykjalin replied:
Fixed in d94f6574df88a7cc160908b77252e27aa71a59a9

Commit:
sync changes made in blocks when payment request is started

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,356 @@
+				if ( result ) {
+					setPaymentRequest( pr );
+					setPaymentRequestType(
+						result.applePay ? 'apple_pay' : 'payment_request_api'
```
We have 3 options, `googlePay` is also allowed here. We need to add it [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/includes/class-wc-stripe-blocks-support.php#L324-L334) and [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/includes/payment-methods/class-wc-stripe-payment-request.php#L403-L421) as well.

ricardo replied:
Same is done in WCPay:

https://github.com/Automattic/woocommerce-payments/blob/4dd975f9aacd589a8a28c60e2672e489cb51eded/includes/class-wc-payments-payment-request-button-handler.php#L364-L383

reykjalin replied:
Oooh, great catch! ❤️

ricardo replied:
I'd also rename "Chrome Payment Request" to just "Payment Request".

This feature is available in Edge, Opera and Brave. It's misleading to name it "Chrome Payment Request".

reykjalin replied:
Both fixed in 9b173d8122c99c26ba3040d75937834c4f7e1ded.

Commit:
sync shipping address changes in PRBs to Blocks

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,356 @@
+			return;
+		}
+
+		getCartDetails().then( ( cart ) => {
```
I might be missing something, but what happens when you update the cart items?

You may need to reset the `paymentRequest` object when the cart changes by using something like:

```JavaScript
useEffect( () => {
    setPaymentRequest( null );
}, [
    // These are props from the cart block.    
    shippingData.needsShipping,
    billing.cartTotal,
    billing.cartTotalItems,
] );
```

reykjalin replied:
This is really only run once the site loads and the payment request is created. When you then click the payment request button, the payment request is updated with new information from the cart 🙂 

I think that logic wasn't pushed to GitHub when you were reviewing this, so you probably need to take another look to see the new onClick handler: 

- [onClick](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7476041531065b80dacb9689cae7876196e12fc9/client/blocks/payment-request/hooks.js#L344-L378) handler; updates the payment request once the PRB is clicked 🙂

ricardo replied:
I see. But this is still necessary, because `updatePaymentRequest` can't update the `requiresShipping` prop.

Taking into account the change of `shippingData.needsShipping` and resetting the paymentRequest object, fixes the problem when you remove all shippable products from the cart, leaving only a digital one.

Edit: Since the onClick is handling the updated labels and totals, something like this as a separate hook may fix the shipping issue:

```JavaScript
useEffect( () => {
    setPaymentRequest( null );
}, [ shippingData.needsShipping ] );
```

reykjalin replied:
Ah yea, I forgot to think about that!

I've added this effect to the `usePaymentRequest` hook (the one that creates the payment request) in 87940a9cc65f12f208319fbec3a0e226b8319fc4.
I decided not to create a new hook since this effect seems like a necessary function of managing the payment request. Let me know if you think it should be its own hook 🤔

ricardo replied:
No, it's better the way you did it.

Commit:
sync shipping method changes from PRBs to blocks

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,356 @@
+					} );
+				} );
+			}
+		);
```
These event handlers are the same for the shortcode. They could be reused if we simply export them as separate functions.

I like the approach in WCPay:

```JavaScript
paymentRequest?.on( 'shippingaddresschange', ( event ) =>
    shippingAddressChangeHandler( api, event )
);

paymentRequest?.on( 'shippingoptionchange', ( event ) =>
    shippingOptionChangeHandler( api, event )
);

// ...
```

This prevents a lot of duplication in the shortcode.

But I'm in favor of having better names and separation like you're proposing for these hooks like 
`usePaymentRequest`, `useEventHandlers` instead of `useInitialization`.

reykjalin replied:
That's an excellent observation! I'll move the event handlers into their own module so we can reuse them if/when we move the shortcode implementation to the webpack based build system 🎉

reykjalin replied:
The handlers are in their own separate module as of a6449575a618dfcc82142c8697750682f166d02f.

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,356 @@
+		 * @return {Object|undefined} The object containing `type`, `clientSecret`, and
+		 *                            `redirectUrl`. Undefined if no partails embedded in the url.
+		 */
+		const getRedirectUrlPartials = ( url ) => {
+			const partials = url.match( /^#?confirm-(pi|si)-([^:]+):(.+)$/ );
+
+			if ( ! partials || partials.length < 4 ) {
+				return undefined;
+			}
+
+			const type = partials[ 1 ];
+			const clientSecret = partials[ 2 ];
+			const redirectUrl = decodeURIComponent( partials[ 3 ] );
+
+			return {
+				type,
+				clientSecret,
+				redirectUrl,
+			};
+		};
```
I think the intent confirmation logic would be better outside this hook.

It could be part of the API, as it's done in WCPay. Or somewhere else if the API is only intended for AJAX requests here.

The benefits of this would be being able to use the same logic for credit-card and shortcode PRB 3DS confirmation.

reykjalin replied:
Yeah, I didn't really want this all to be in the hook, I just wasn't sure what to do with it all.

You're right though; I should put this somewhere else so we can re-use this logic in the future.

> Or somewhere else if the API is only intended for AJAX requests here.

Parts of the functionality here should be re-usable across PRBs vs. CC and Blocks vs. shortcode. I'll rethink this and see if I can move the functions to a better place that allows them to be easily reused later 🤔

reykjalin replied:
This is now part of the `event-handlers` module, but should eventually be moved outside of that module to somewhere where both the Credit Card and PRB checkout code can access and use it.

Do you think we should do that as part of this PR?

ricardo replied:
No. I think we can open another PR for that.

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,356 @@
+
+		// Update the payment request with new billing information.
+		if ( paymentRequest ) {
+			updatePaymentRequest( {
+				paymentRequest,
+				total: billing.cartTotal,
+				currencyCode: billing.currency.code.toLowerCase(),
+				cartTotalItems: billing.cartTotalItems,
+			} );
```
`usePaymentRequest` is using `getCartDetails()` to retrieve totals and display items.

There's this inconsistency for the labels when retrieving from the `billing` prop or from the AJAX request.

I don't have an idea on how to fix this yet, but perhaps we can still use the `billing` and `shippingData` prop in `usePaymentRequest` so we save one unnecessary request here.

reykjalin replied:
Hmmm, I'd like the labels to be the same for shortcode and Blocks 🤔 
That's why _ideally_ we'd use the AJAX API to set the label, but you're right; when we populate with information from the block that changes the label.

I'll look into this. Nice catch!

reykjalin replied:
Hmm, I'm honestly not sure how to fix this. I think we have to do 1 of 3 things:

1. Use `getCartDetails()` for everything in blocks and spend the time to make sure the shortcode and blocks PRBs populate the payment request with the same data
1. Use `getCartDetails()` for everything in blocks and accept the inconsistencies.
1. Use the AJAX API for everything (both blocks and shortcode) and accept that we have to make a request for the cart each time we call `updatePaymentRequest`.

The downside of (1) is extra maintenance — changes in one place will need changes in another place and those are really difficult to remember.

The downside of (2) is the inconsistency — The payment requests dialogs won't have the exact same information. When we start customizing them (e.g. when merchants enable the hook that adds an itemized list of things being purchased) we'll have to implement that twice, possibly in 2 different ways.

The downside of (3) is extra network requests.

I think (3) is the best option. It's a small network request we're trading for what would be a significant maintenance burden.

ricardo replied:
I'm a bit confused about these options because they sound like the same thing to me 😆

By using `getCartDetails()` for everything we're basically doing option 1 and 2 at the same time, and `getCartDetails()` is making the extra request, which is part of the AJAX API (3).

But I agree, if we use `getCartDetails()` for everything, including `updatePaymentRequest`, we will have more consistency with the labels between blocks and shortcode, but the downside is the extra request.

I think it's fine to use it in both places.

reykjalin replied:
> and `getCartDetails()` is making the extra request

ugh, sorry, I meant "use the Blocks data (i.e. front-end data) to populate the payment request" instead of `getCartDetails()` 🤦 

To clear this up a bit more: The options I meant to propose:

1. Use the [front-end data](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/client/blocks/payment-request/use-initialization.js#L70-L77) to populate the payment request and fix up the labels so Shortcode and Blocks Payment Requests look the same.
2. Use the [front-end data](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/client/blocks/payment-request/use-initialization.js#L70-L77) to populate the payment request and accept any inconsistencies Shortcode and Blocks Payment Requests.
3. Use `getCartDetails()` (the AJAX API) to populate both Shortcode and Blocks Payment Requests.

The downsides list I added above is based on this ^ list, not the one in my original comment.

Sorry for the misunderstanding!

> But I agree, if we use getCartDetails() for everything, including updatePaymentRequest, we will have more consistency with the labels between blocks and shortcode, but the downside is the extra request.
>
> I think it's fine to use it in both places.

Just to be clear; you mean it's fine to use `getCartDetails()` in both places, yes? 🙂

reykjalin replied:
Ok, so I just made the change to use `getCartDetails()` for both creating the payment request, as well as updating it in 17af3684fe75db46fab8aa3b8d29a8909692f78a. I had to change the `onClick` handler a bit to accomodate the network request, but I honestly think it makes more sense the way it is now ☺️

ricardo replied:
> Just to be clear; you mean it's fine to use getCartDetails() in both places, yes? 🙂

Yes. 👍 

> I had to change the onClick handler a bit to accomodate the network request, but I honestly think it makes more sense the way it is now 

I see a few problems with the async request on click:
- A request is made even when nothing has changed.
- There's a delay for the response when you click the button.
- There's a warning `Do not call show() ...` when you click the button.

I think a better approach would be to reset the button only when `needsShipping` or `billing` changes.

You could simply do this inside `usePaymentRequest`, and discard the `paymentRequest.update` logic (billing needs to be passed to `usePaymentRequest`):

```
useEffect( () => {
	setPaymentRequest( null );
}, [ needsShipping, billing.cartTotal, billing.cartTotalItems ] );
```

You still need to call `setExpressPaymentError` and `onClick` inside the `useOnClickHandler` hook, but `updatePaymentRequestWithCart` is not necessary.

The suggestion above solves the 3 issues mentioned by calling `getCartDetails()` only when necessary.

ricardo replied:
🤔 I was wondering just now - can the currency code change from the cart block? I've seen the previous implementation use `billing.currency.code.toLowerCase()`.

If so, we'd need to add `billing.currency.code` to the above suggestion, and somehow return the cart currency code from `getCartDetails()`.

reykjalin replied:
> I think a better approach would be to reset the button only when needsShipping or billing changes.

Yeah, that makes sense.

Fixed in 8924d4c16e12c97b9d9b23fcea5e24d189067948.

> There's a warning Do not call show() ... when you click the button.

To be fair, that happens even though I'm calling the `preventDefault()` function provided by Stripe 🙂

Commit:
Add cancellation handler to the payment request

ricardo reviewed (approved):
This is coming along nicely 👏  

This not a final review. Just some comments I had underway before your latest commits.

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+			( evt ) => {
+				const { shippingAddress } = evt;
+
+				// Update the block shipping address, just in case the payment request
+				// is cancelled.
+				setShippingAddress(
+					normalizeShippingAddressForCheckout( shippingAddress )
+				);
```
Is there a reason for this `setShippingAddress`?

![image](https://user-images.githubusercontent.com/5509901/117502131-754d1900-af55-11eb-8d7c-b2053950c339.png)

This ends up making a second request to the Gutenberg endpoint, which leads to an error in those edge cases where the ZIP code is redacted or the state is not what WooCommerce is expecting (e.g Chinese provinces).

IMO the address choosing inside the payment request flow can be separate from the cart block shipping calculation. I don't know how we could manage to do it otherwise, given these inconsistencies between the address validation in the Payment Request API and what WooCommerce expects.

reykjalin replied:
Yeah that's an excellent point and not something I considered!
I agree, it's better (at least for now) to leave these 2 separated.

Removed in f47a87bf9a53ece85fa8417ec625e444f289edbc

ricardo commented:
![image](https://user-images.githubusercontent.com/5509901/117502449-e8ef2600-af55-11eb-98ae-21f6007c4536.png)

WooCommerce wraps the error messages with `wc_add_notice` and returns these HTML error notices as responses. 🤷🏻‍♂️

I have got around this in WCPay by creating [this function](https://github.com/Automattic/woocommerce-payments/blob/7aba40a8404ebad9fd877a8d9e753525d612e820/client/payment-request/utils/utils.js#L129-L133) to extract the error message from these notices, so you could do something like `setExpressPaymentError( getErrorMessageFromNotice( error ) )`.

🤷🏻‍♂️ I don't like it, but it should work.

ricardo commented on client/api/index.js:
```diff
@@ -0,0 +1,166 @@
+			name?.split( ' ' )?.slice( 0, 1 )?.join( ' ' ) ?? '',
+		billing_last_name: name?.split( ' ' )?.slice( 1 )?.join( ' ' ) ?? '',
+		billing_company: '',
+		billing_email: email ?? sourceEvent.payerEmail,
```
We should probably use `sourceEvent?.payerEmail` here as well.

reykjalin replied:
Ooh, this is a great catch! I missed that, thank you!

Fixed in 98ee96a7d435887036d5c67aff531d06469b2a77

ricardo commented on client/api/index.js:
```diff
@@ -0,0 +1,166 @@
+	const data = {
+		security: wc_stripe_payment_request_params.nonce.update_shipping,
+		shipping_method: [ shippingOption.id ],
+		payment_request_type: paymentRequestType,
```
`payment_request_type` is not used in `ajax_update_shipping_method`. It's only used in `postal_code_validation` in `get_shipping_options`.

reykjalin replied:
Nice catch!

Removed the parameter in cd8308bc40410375955a27a1284c17812251f57d

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+import {
+	updatePaymentRequest,
+	normalizeShippingAddressForCheckout,
+} from '../stripe-utils';
```
It makes sense to put `updatePaymentRequest` and `createPaymentRequest` into the same group/file I think.

reykjalin replied:
Yeah, you're right.

Moved `createPaymentRequest` to `stripe-utils` in 1832ea792a7288f9a965b7bbeaed95906f7e960f 🙂

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+		);
+
+		return () => {
+			// Need to use `?.` here in case shippingAddressHandler is null.
```
Comment needs to be updated to `shippingAddressUpdateHandler`.

reykjalin replied:
Fixed in 64c6e43e431c287857ec2d31cb1589a2101c6db4

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+) => {
+	useEffect( () => {
+		// Need to use `?.` here in case shippingAddressHandler is null.
+		const shippingOptionChangeHandler = paymentRequest?.on(
```
Comment should be `... in case paymentRequest is null.`

reykjalin replied:
Fixed in 64c6e43e431c287857ec2d31cb1589a2101c6db4

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+				const { shippingOption } = evt;
+
+				// Update the selected rates in the cart or checkout block.
+				setSelectedRates( shippingOption?.id );
```
Same as in `setShippingAddress`. We might not want to use `setSelectedRates` to mirror the selection in the payment request dialog, because this has two problems:
1. It may not work for those edge cases which require address normalization.
2. It adds an additional request.

reykjalin replied:
That's a really good point that I didn't consider, thank you! ❤️ 

Removed in f47a87bf9a53ece85fa8417ec625e444f289edbc

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,392 @@
+							"Sorry, we're not accepting prepaid cards at this time.",
+							'woocommerce-gateway-stripe'
+						)
+					);
```
You could also use `wc_stripe_payment_request_params.i18n.no_prepaid_card` here, so we don't need to localize this script. I guess this is the only place where we use i18n for PRB.

reykjalin replied:
I like this suggestion! Using the pre-existing one will be better for consistency as well 🎉 

Fixed in c69c46150a5a7cbb7b044abfc8128754459f4181.

ricardo reviewed (changes requested):
This tests well and the code looks good. 👍 

Kudos for removing the previous hard-to-parse hooks. It's a lot simpler now. 👏

You can ignore my comments about reusing the same code for 3DS and shortcode event handlers. We can do it at a later time.

I only left a couple comments about the error notice and changing the state of the cart block from the PRB event handlers, which I think should be fixed now.

Commit:
fix wrong variable names in comments

Commit:
Remove sync from PRB options to Block checkout

@ricardo made a good point; because the validation done for addresses
is different for Payment Requests and WooCommerce this can cause issues
where an address is valid for PRBs, but not for WC.

These also caused an extra request being sent to the server which slows
the flow down.

Commit:
add missing ?. operator

Commit:
Remove unnecessary parameter from API call

The `payment_request_type` paremeter is not required to update the
chosen shipping method, so this commit removes the need for that
parameter.

Commit:
use pre-existing translation string instead of creating a new one

Commit:
Move createPaymentRequest function to stripe-utils

Commit:
Ensure PRBs react to changes in shipping requirements

Commit:
move order data normalization to normalization module

Commit:
move payment request event handlers to separate module

Commit:
Add GPay Payment Request title

Commit:
Show a better error message when PRB checkout requires an account

reykjalin commented:
> WooCommerce wraps the error messages with `wc_add_notice` and returns these HTML error notices as responses. 🤷🏻‍♂️
> 
> I have got around this in WCPay by creating [this function](https://github.com/Automattic/woocommerce-payments/blob/7aba40a8404ebad9fd877a8d9e753525d612e820/client/payment-request/utils/utils.js#L129-L133) to extract the error message from these notices, so you could do something like `setExpressPaymentError( getErrorMessageFromNotice( error ) )`.
> 
> 🤷🏻‍♂️ I don't like it, but it should work.

Ooof, this is really bad 😬 

![image](https://user-images.githubusercontent.com/13835680/117582937-170a6c80-b0f4-11eb-8960-9bc844523823.png)

The messaging is especially bad, but I think this is a known issue. I've implemented your suggestion in 769e6efe8cc47c440460def4cb414688c467f7bd
, but I think we need to do something more intelligent here. I don't think that should be part of this PR though.

cc @v18 and @LevinMedia 

I also feel like we talked about this before 🤔 . The question we need to answer is:

"How do we tell people they need to create an account instead of showing them the unhelpful error message we currently do, when paying with a Payment Request without creating an account?"

I think the solution in WCPay was "don't show Payment Request Buttons if the customer isn't logged in?".

**Note:** To see this issue the merchant must turn off automatic account creation in their WooCommerce Settings.

ricardo commented:
@reykjalin  Yes, this is a known issue. I was just pointing out to the HTML in the error notice. Any WooCommerce errors would appear like that.

> "How do we tell people they need to create an account instead of showing them the unhelpful error message we currently do, when paying with a Payment Request without creating an account?"

For this, last week we decided on a new approach. I'll be working on it - p1620311692228500-slack-CGGCLBN58

Commit:
Update payment request on click with data from AJAX API

ricardo commented on client/api/index.js:
```diff
@@ -0,0 +1,86 @@
+				: address.addressLine[ 1 ],
+		payment_request_type: paymentRequestType,
+		is_product_page: wc_stripe_payment_request_params.is_product_page,
+	};
```
Not a blocker and feel free to ignore, but like `normalizeOrderData`, I'd count this as normalization as well and include in a separate function.

reykjalin replied:
Fixed in f829c9e9eca7a4437409aab1858e06012012ccc1

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,214 @@
+export const useShippingAddressUpdateHandler = (
+	paymentRequest,
+	paymentRequestType,
+	setShippingAddress
```
`setShippingAddress` doesn't seem to be used anymore, I guess we can remove it.

reykjalin replied:
Fixed in 4d23f8c50c2a632f93d3f5100c7fdf6b1e3ca896.

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,214 @@
+ *
+ * @param {Object} paymentRequest - The payment request object.
+ * @param {string} paymentRequestType - The payment request type.
+ * @param {Function} setSelectedRates - A function used to set the selected shipping method in the
```
`setSelectedRates` is also not being used. We can remove it.

reykjalin replied:
Fixed in 4d23f8c50c2a632f93d3f5100c7fdf6b1e3ca896.

ricardo commented on client/blocks/payment-request/hooks.js:
```diff
@@ -0,0 +1,214 @@
+		return () => {
+			paymentMethodUpdateHandler?.removeAllListeners();
+		};
+	}, [ stripe, paymentRequest, paymentRequestType, setExpressPaymentError ] );
```
`paymentRequestType` is set always along with `paymentRequest`. So perhaps we could reduce the noise and remove `paymentRequestType` from here and from `useShippingAddressUpdateHandler` and `useShippingOptionChangeHandler`.

reykjalin replied:
Personally I think it's better to be explicit here. If we ever make a change and update `payementRequestType` in a different place from `paymentRequest` I think it'll be difficult to remember to update these.

That said, ESLint also warns if you don't include _everything_ you use inside `useEffect`, and I don't think we should be ignoring that warning.

ricardo commented on client/blocks/payment-request/payment-request-express.js:
```diff
@@ -98,33 +107,17 @@ const PaymentRequestExpressComponent = ( {
 	const isCustom = wc_stripe_payment_request_params.button.is_custom;
 
-	if ( ! canMakePayment || ! paymentRequest ) {
+	if ( ! paymentRequest ) {
```
I'm not sure if we can remove `canMakePayment` from here. Stripe returns a `canPay` prop for some reason, but I didn't dig into how that works.

For reference:
https://github.com/Automattic/woocommerce-payments/blob/2d7f64ef5e7817a39fb2c4be068a94b6b16dd608/client/payment-request/blocks/use-initialization.js#L61

reykjalin replied:
Apparently, the `.canPay` prop [is not a thing](https://stripe.com/docs/js/payment_request/can_make_payment). I don't know where that comes from originally 🤔 
Or at the very least it's not a _documented_ thing. I don't think we should be relying on it 🤷

ricardo replied:
Ah right! It's not returned by Stripe, but from `canDoPaymentRequest`. My apologies. Let's not worry about it 👍

ricardo commented on client/blocks/payment-request/payment-request-express.js:
```diff
@@ -98,33 +107,17 @@ const PaymentRequestExpressComponent = ( {
-				} }
-			/>
-		);
+		return <CustomButton onButtonClicked={ onPaymentRequestButtonClick } />;
```
We still need a special callback for `CustomButton` and `GooglePayButton`, because `PaymentRequestButtonElement` calls `paymentRequest.show()`, so in that case it gets called twice, leading to the following warning: 

```
Do not call show() yourself if you are using the paymentRequestButton Element. The Element handles showing the payment sheet.
```

reykjalin replied:
This wasn't _technically_ a problem, but was removed and fixed anyway in https://github.com/woocommerce/woocommerce-gateway-stripe/commit/8924d4c16e12c97b9d9b23fcea5e24d189067948 🙂

ricardo commented on client/blocks/stripe-utils/normalize.js:
```diff
@@ -251,6 +321,7 @@ export {
+	normalizeOrderData,
 	getBillingData,
 	getPaymentMethodData,
 	getShippingData,
```
- `normalizeShippingOptions`
- `normalizeShippingAddressForCheckout`
- `normalizeShippingOptionSelectionsForCheckout`
- `getBillingData`
- `getPaymentMethodData`
- `getShippingData`

Don't seem to be used anymore. I guess we can remove them.

reykjalin replied:
Amazing catch! I went through more of the codebase and removed a bunch of unused functions in 4d23f8c50c2a632f93d3f5100c7fdf6b1e3ca896.

ricardo commented on client/blocks/stripe-utils/utils.js:
```diff
@@ -131,6 +163,21 @@ const updatePaymentRequest = ( {
+ * @param {StripePaymentRequest} paymentRequest - The Stripe Payment Request object.
+ * @param {Object} cart - The cart data response from the store's AJAX API.
+ */
+const updatePaymentRequestWithCart = ( paymentRequest, cart ) => {
+	paymentRequest.update( {
+		total: cart.order_data.total,
+		currency: cart.order_data.currency,
+		displayItems: cart.order_data.displayItems,
+	} );
+};
```
`updatePaymentRequestWithCart` and `updatePaymentRequest` can be removed if you implement [this suggestion](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1545#discussion_r629821055).

There are some other unnecessary functions:
- `getTotalPaymentItem`
- `getPaymentRequest`
- `canDoPaymentRequest`
- `pluckAddress`

reykjalin replied:
Fixed in 4d23f8c50c2a632f93d3f5100c7fdf6b1e3ca896.

ricardo reviewed (changes requested):
Nice work 👍 👏 

Just a few comments and some unused code I didn't notice before.

v18 commented:
> I think we need to do something more intelligent here. I don't think that should be part of this PR though

Do you mean for extracting the error message from the HTML? Or for the account required problem?

For the account required - let's not show the PRBs when an account is required - same as we do for the shortcode. And then we'll fix the real issue after we hear back from David about how the flow should work.

Happy for that to be addressed as another PR.

ricardo commented:
@v18 The account required issue is fixed in trunk. It was fixed a while ago, this branch is a bit behind, but it should work when rebased.

@reykjalin For the blocks to fully work with that you need to adapt two things. Sorry for pointing out just now. This PR has no diff on these files, so I forgot to mention it:

1. On `client/blocks/index.js` remove the `getStripeServerData().allowPaymentRequest` check. This prevents the payment method from being registered in the editor context. You can remove the `allowPaymentRequest` variable from the back-end too ([WCPay reference](https://github.com/Automattic/woocommerce-payments/blob/3ee77525a525e46caffe087c7d4754cc686c6d69/client/checkout/blocks/index.js#L51)).
2. On `client/blocks/payment-request/index.js`, make sure `canMakePayment` returns true only if `wc_stripe_payment_request_params` is defined ([WCPay reference](https://github.com/Automattic/woocommerce-payments/blob/3ee77525a525e46caffe087c7d4754cc686c6d69/client/payment-request/blocks/index.js#L15-L26)). This will prevent the button from showing up if it's not supported.

To be honest, I think [`paymentRequestAvailable`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/17af3684fe75db46fab8aa3b8d29a8909692f78a/client/blocks/payment-request/index.js#L24-L54) can be dismissed and simplified like in WCPay, first because the credit card method doesn't take into account any of those Stripe error checks, and second because the shortcode doesn't check for any of that too. It doesn't really click to me to perform a test payment to confirm if the method is available.

Commit:
reset payment request on cart change instead of updating it

Commit:
remove a bunch of unused functions

Commit:
move normalization functionality to the normalization module

Commit:
Make canMakePayment check simpler

Instead of trying to make a fake payment to determine whether a payment
request can be made we instead just check whether the Stripe JS object
and the Stripe Pluguin Payment Request options are available and trust
that payment requests can be made if both are present.

reykjalin commented:
> Sorry for pointing out just now. This PR has no diff on these files, so I forgot to mention it:

No worries! I'm just happy you thought of this now 😅 

> On `client/blocks/index.js` remove the `getStripeServerData().allowPaymentRequest check`.
> [...]
> `client/blocks/payment-request/index.js`, make sure `canMakePayment` returns true only if `wc_stripe_payment_request_params` is defined.

Done in 2dafe4260aac35a839edd49910ebd42f4203f4c1.

> To be honest, I think `paymentRequestAvailable` can be dismissed and simplified like in WCPay, first because the credit card method doesn't take into account any of those Stripe error checks, and second because the shortcode doesn't check for any of that too. It doesn't really click to me to perform a test payment to confirm if the method is available.

Agreed 💯 
Removed `paymentRequestAvailable` in 2dafe4260aac35a839edd49910ebd42f4203f4c1 🙂

ricardo reviewed (approved):
Changes look good to me 👍 
Thanks for addressing all issues so quickly. ⏩ 

I'm approving the changes, but I found one issue with the payment method title under **WooCommerce > Orders > Edit order**. It says "Payment via Credit Card (Stripe)" instead of "Apple Pay", "Google Pay" or "Payment Request".

I don't know where the bug is, but this was probably originated in this branch. I tested with `feature/support-block-checkout` and it works fine.

Commit:
Fix title matching for orders paid via PRBs

This makes sure the payment method title is correctly represented in the
order information. After changing 'Chrome Payment Request' to just
'Payment Request' I forgot to change the matching done in this commit to
properly display the right payment method title.
This has now been fixed.

reykjalin commented:
@ricardo sorry about that! Turns out we did some specific filtering to correctly display the right title and I forgot to update that following the changes in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1545/commits/9b173d8122c99c26ba3040d75937834c4f7e1ded.

I've fixed the issue in 0ed575dbd25bf8604bee86cfada781921d186e82 and it should apply for any previous orders as well. 👍 

I think everything should be in order now 🙂

Commit:
add Google Pay title filter

ricardo reviewed (approved):
Tests well & looks good! 🚢

Commit:
Update translation strings
